src/math/dd/dd_bdd.cpp

```diff
@@ -896,87 +911,54 @@ namespace dd {
     std::ostream& operator<<(std::ostream& out, bdd const& b) { return b.display(out); }
 
 
-    // NSB code review: 
-    // this function should be removed and replaced by functionality where the
-    // client maintains what are the variables.
-    bdd bdd_manager::mk_int(rational const& val, unsigned w) {
-        bdd b = mk_true();
-        for (unsigned k = w; k-- > 0;)
-            b &= val.get_bit(k) ? mk_var(k) : mk_nvar(k);
-        return b;
-    }
-
-    bool bdd_manager::contains_int(BDD b, rational const& val, unsigned w) {
+    bool bdd_manager::contains_num(BDD b, rational const& val, unsigned_vector const& bits) {
+        DEBUG_CODE(for (unsigned i = 1; i < bits.size(); ++i) { SASSERT(bits[i-1] < bits[i]); });
+        unsigned var_idx = bits.size();
         while (!is_const(b)) {
-            unsigned const var = m_level2var[level(b)];
-            if (var >= w)
-                b = lo(b);
-            else
-                b = val.get_bit(var) ? hi(b) : lo(b);
+            VERIFY(var_idx-- > 0);
+            SASSERT(var(b) <= bits[var_idx]);
+            while (var(b) < bits[var_idx]) {
+                VERIFY(var_idx-- > 0);
+            }
+            SASSERT(var(b) == bits[var_idx]);
+            b = val.get_bit(var_idx) ? hi(b) : lo(b);
         }
         return is_true(b);
     }
 
-    find_int_t bdd_manager::find_int(BDD b, unsigned w, rational& val) {
+    find_result bdd_manager::find_num(BDD b, unsigned_vector bits, rational& val) {
```

Comment:
Instead of all of this inside the bdd manager, the polysat solver object could maintain a layer (a class) that maintains the association of variables. Pushing this functionality into bdd_manager is so far awkard. For a BDD native version you would basically want to pass in a function that maps variables to power of two. Together with some accounting for how many variables were fixed.